### PR TITLE
feat: e2e test using POM pattern

### LINKED ISSUE #24: Write unit & integration tests for contributions form
## Description

Increase code coverage of the form components used in the New Contribution page.

This issue is demonstrated in this talk:

* **Software Web Testing**: [_Pengujian Perangkat Lunak Web_](https://www.facebook.com/rr17x/posts/pfbid068qzcd2tvVUKVkLrLwApbktM5hsWTrcFJuqjhybRNh2BD7YdL65Vrgy3dAD4oudLl?__cft__[0]=AZXlBEILFOeeXaOgz0LgIbAT0W0zVCwZj_flRa72RbxGNMqt4EZTl12fq0dOjg4fiWN26AwwNs1WRbldLqJaD_7s9AbyMgo6LUgsJXOc7i3R1YSyZDpOM1ISMpP5KxM4LbuptBli4ak4sIkfdUedw_jU&__tn__=%2CO%2CP-R).

## Narrative

- **As a** public visitor
- **I want** to submit an activity to the contribution form
- **so that** I can get my activity reviewed by the curators.

### Scenario 1: Successful form submission

- **Given** I visit `/contributions/new` page
- **and** I fill in the form correctly
- **when** I submit the form
- **then** I should be redirected to a `/contributions/:activityId` route
- **and** I can see my submission has been recorded correctly.

### Client

- [x] Extract the contribution form component from `/contributions/new.tsx` route.
  - [x] Ensure that the end-to-end test scenarios are not broken after the component extraction.
- [ ] Write unit & integration tests for the contribution form component, which should also cover the `form-elements.tsx` file.
  - [x] Assert form submission
  - [ ] Assert styling
     - [ ] Inline snapshot
  - [x] Make the tests easier to read by utilising [Faker](https://fakerjs.dev) and [test-data-bot](https://github.com/jackfranklin/test-data-bot)
  - [ ] Assert web accessibility
     - [x] ARIA
     - [ ] Focus Management
     - [ ] Test and improve VoiceOver Navigation
        To activate/deactivate VoiceOver, press `Fn + CMD + F5`. (On a Mac desktop keyboard, you can omit the `Fn` key.)

### Server

- [ ] Extract the FormData transformation to Activity entity
- [ ] Test the FormData transformation
- [ ] Refactor the imperative data transformation into [functional paradigm](https://www.facebook.com/gifa.eriyanto/posts/pfbid0rEKy8hR7Uq4rKLNzzRenCiGrwzRJXpXrJUoJo1kgpHq1f4h9fEuWGvNS4z8rUrMml?__tn__=%2CO*F)
- [ ] Use [zod](https://zod.dev) for data validation

### PR DESCRIPTION
**Pull request by depapp**
Closes https://github.com/zainfathoni/senar.ai/issues/24#issuecomment-1276887934 

## Description

refactor e2e test to use Page-Object-Model pattern.

## Current Tasks

- [x] `homepage` test

### CODE REVIEW

File: playwright.config.ts [hunk 1]
```diff
@@ -40,7 +40,7 @@ const config: PlaywrightTestConfig<Fixtures> = {
     baseURL: 'http://localhost:3000',
 
     /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
-    trace: 'on-first-retry',
+    trace: 'retain-on-failure',
```
Comment (zainfathoni) [inline playwright.config.ts:43]: Nice, thanks! 👍

File: e2e/pages/home.page.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { Page, Locator } from '@playwright/test';
+
+export class HomePage {
+  readonly page: Page;
+  readonly activityCard: Locator;
+  readonly paudActivity: Locator;
+
+  constructor(page: Page) {
+    this.page = page;
+    this.activityCard = page.locator('h3.text-lg.font-medium');
```
Comment (zainfathoni) [inline e2e/pages/home.page.ts]: I am a proponent of writing tests that imitate how the users use the app.
In this case, the user won't look for any `h3` elements with `text-lg font-medium` class.
So let's avoid this kind of selector.

Instead, you can use Playwright's [new APIs inspired by Testing Library](https://twitter.com/diegohaz/status/1578567655493193728). But first, you need to upgrade the Playwright version to [1.27.0](https://github.com/microsoft/playwright/releases/tag/v1.27.0) to use it.

File: e2e/pages/home.page.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    this.paudActivity = page.locator('text=PAUDPendidikan Anak Usia Dini >> span');
```
Comment (zainfathoni) [inline e2e/pages/home.page.ts]: Same as above. The user won't try to find any `span` in the document.

**Review by zainfathoni — CHANGES_REQUESTED** [review]
Thanks for refactoring it. I like the way it's organized now. 💯 
However, I have some objections in the way we query the elements. 😅 

Quoting from [Testing Library](https://testing-library.com) slogan,
> The more your tests resemble the way your software is used, 
the more confidence they can give you.

So please update the queries to use my initial queries from [playwright-testing-library](https://www.npmjs.com/package/playwright-testing-library), or you can also experiment with Playwright's [new APIs inspired by Testing Library](https://twitter.com/diegohaz/status/1578567655493193728). But first, you need to upgrade the Playwright version to [1.27.0](https://github.com/microsoft/playwright/releases/tag/v1.27.0) to use it.

File: e2e/pages/home.page.ts (hunk 2 above)
Comment (depapp) [reply e2e/pages/home.page.ts]: noted with thanks @zainfathoni 
I'll try to use this one [new APIs inspired by Testing Library](https://twitter.com/diegohaz/status/1578567655493193728)

**Comment by depapp** [comment]
could you please help to recheck @zainfathoni , thanks!
image

**Review by zainfathoni — APPROVED** [review]
LGTM 💯
Thanks for making the changes accordingly, @depapp! 🙏

**Comment by zainfathoni** [comment]
@all-contributors add @depapp for test

**Comment by zainfathoni** [comment]
@all-contributors add @depapp for test

**Pull request merged**